### PR TITLE
URIRef comparison with other types fixed, see #773

### PR DESCRIPTION
**Pull request by luav**
URIRef comparison with other types fixed, see #773 (caused invalid comparison of URIRef with str on Python 3)

### CODE REVIEW

**Comment by luav** [comment]
Implemented matching of `Identifier` subclasses to the same types, to the base type `Identifier` and to the raw python strings. Note that different subclasses of `Identifier` (`Literal` vs `BNode`, etc.) remained incomparable between each other.
Initial version where `Identifier` subclasses  like `URIRef` could be extended with python `str`, but silently failed to compare to `str` was confusing and unexpected.

Basic tests are passed now, except tests in `test_dawg.py`. I have no idea about that tests and don't want to spend my time for it. Feel free to complete the integration!

**Comment by gromgull** [comment]
Hi @luav - Thanks for you input!

Unfortunately, `test_dawg` is sort of "the most important test". It tests the SPARQL unit tests from the data-access group. RDFLib implements SPARQL semantics by directly comparing the RDFLib terms, for SPARQL it is crucial that the URI `<http://example.org/a>` is NOT equal to the string "<http://example.org/a>"`

In fact, this is pretty much crucial for all of RDFLib - I don't see why you would want it to be otherwise? 

The real issue here is probably that we inherit from unicode/str - leading people like you to believe they are similar, but then they are not. We have discussed removing unicode from the class hierarchy before, but I can't find the pointer right now. 

So unless there are some VERY good arguments I am missing, I am "wont fix" on this one. Sorry!

**Comment by luav** [comment]
Hi @gromgull ,
you are right that `http://example.org/a` != URI `<http://example.org/a>` but I'm talking about another comparison, comparison of the str `<http://example.org/a>` vs URI `<http://example.org/a>`.
Please, consider the following example:
```python
import rdflib
from rdflib.namespace import RDF
# print(RDF.type) yields: http://www.w3.org/1999/02/22-rdf-syntax-ns#type
typeProp = '<' + RDF.type + '>'  # N3/quad type property
prop = '<http://www.w3.org/1999/02/22-rdf-syntax-ns#type>'

eq1 = typeProp == prop  # False
eq2 = len(prop) == len(typeProp) and not prop.find(typeProp)  # find returns 0 for the match
assert eq1 == eq2, '{} vs {}'.format(eq1, eq2)
```
yields
```sh
>>> assert eq1 == eq2, '{} vs {}'.format(eq1, eq2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AssertionError: False vs True
```
The representation of URI was ambiguous and misleading before the suggested fix. If your show the content of URI without the `<`, `>`  but semantically assume presence of the URI brackets then the modifying operations ('+', etc.) should behave differently, which would not be a Pythonic way of doing things and extremely hard to be properly controlled and efficiently implemented.

Basically, to avoid any concerns and be consistent, the URI either should return str URI (including `<` '>') or should support consistently all possible operations (assignment, comparison, modification, etc.).

**Comment by white-gecko** [comment]
Hi @luav ,
thank you very much for all the work you put into the code and also in making an argument. I can understand that the current behavior of the rdflib in combination with the python datatypes is not always intuitive. As changing this behavior would break the current API we will definitely not introduce these changes in the 5.0.0 release. For the 6.0.0 release we want to drop Python 2 support and have some major rework. We also think about doing something to the relation of RDFLib objects and python data types (@nicholascar I hope I got this one right). I sure this will involve some discussion were I would like to welcome you.
Adding the angle brackets around the IRI would follow the Turtle, NTriples, Trig, NQuads, and SPARQL syntax, which is ok in my eyes, as it would not be easy to serialize individual nodes following JSON-LD or RDF/XML. In addition we would also need to see if we deal with the other RDF terms in the same way and print Literals as `"Literal"@en` instead of `Literal` which would definitely break the current API.
I hope this as well as the earlier comment as well as the comment in #773 depict how complicate it is and that we can not accept this PR at the moment.
I hope it is ok for you that I close this PR and #773 and put them on "out of scope".
